### PR TITLE
Versioning (implementation of functionality)

### PR DESCRIPTION
**Pull request by stingion**

### CODE REVIEW

File: scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ConfigurationService.java [hunk 1]
```diff
@@ -47,7 +48,16 @@
...
+  Mono<List<ReadEntryResponse>> readList(ReadListRequest request);
+
+  /**
+   * Fetch request requires read level permissions to get entry object from the store.
+   *
+   * @param request includes the repository and key of the requested object.
+   * @return json object from the store.
+   */
+  @ServiceMethod
+  Mono<ReadEntryResponse> readEntryHistory(ReadEntryHistoryRequest request);
```
Comment (dmytro-lazebnyi) [inline scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ConfigurationService.java]: Shouldn't it be `Mono<List<ReadEntryResponse>>`?

File: scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ReadEntryHistoryRequest.java [hunk 2]
```diff
@@ -1,7 +1,7 @@
 package io.scalecube.configuration.api;
 
 /** Represents a request o fetch data from a repository. */
-public class FetchRequest implements AccessRequest {
+public class ReadEntryHistoryRequest implements AccessRequest {
```
Comment (dmytro-lazebnyi) [inline scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ReadEntryHistoryRequest.java]: Why `FetchRequest` renamed to `ReadEntryHistoryRequest`?
Shouldn't it be `ReadEntryRequest`?

File: scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ReadEntryHistoryRequest.java [hunk 3]
```diff
@@ -12,31 +12,21 @@
    *
    * @deprecated only for serialization/deserialization.
    */
-  FetchRequest() {}
+  ReadEntryHistoryRequest() {}
 
   /**
-   * Constructs a FetchRequest object.
+   * Constructs a ReadEntryRequest object.
```
Comment (dmytro-lazebnyi) [inline scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ReadEntryHistoryRequest.java]: Since this class is `ReadEntryHistoryRequest`, so java doc should describe the same class name.

File: scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ReadEntryHistoryRequest.java [hunk 4]
```diff
@@ -51,6 +41,7 @@ public Object apiKey() {
 
   @Override
   public String toString() {
-    return "FetchRequest [repository=" + repository + ", key=" + key + ", apiKey=" + apiKey + "]";
+    return "ReadEntryRequest [repository=" + repository + ", key=" + key + ", apiKey=" + apiKey
```
Comment (dmytro-lazebnyi) [inline scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ReadEntryHistoryRequest.java]: Same situation as with Javadoc.

File: scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ReadEntryHistoryResponse.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package io.scalecube.configuration.api;
+
+/**
+ * Represents a response to a key version history request.
+ */
+public class ReadEntryHistoryResponse {
```
Comment (dmytro-lazebnyi) [inline scalecube-configuration-api/src/main/java/io/scalecube/configuration/api/ReadEntryHistoryResponse.java]: Why do we need such a response type?
Isn't enough to use `List<ReadEntryResponse>` as the return type of the service method?

File: scalecube-configuration/src/main/java/io/scalecube/configuration/repository/ConfigurationRepository.java [hunk 6]
```diff
@@ -41,7 +43,17 @@
...
+
+  /**
+   * Update a key in a tenant repository by key.
+   *
+   * @param tenant namespace of the repository.
+   * @param repository name to update key.
+   * @param doc the document to update.
+   * @return Document instance by a given key and version.
+   */
+  Mono<Document> updateEntry(String tenant, String repository, Document doc);
```
Comment (dmytro-lazebnyi) [inline scalecube-configuration/src/main/java/io/scalecube/configuration/repository/ConfigurationRepository.java]: Why do we need to change `Repository` interface for the sake of public API (service API) changes?

File: scalecube-configuration/src/test/java/io/scalecube/configuration/CouchbaseExperimentalTest.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import com.couchbase.client.java.document.json.JsonObject;
+import java.time.Duration;
+import java.util.Random;
+import java.util.UUID;
+import org.junit.Test;
+import org.junit.jupiter.api.Disabled;
+import reactor.core.publisher.Mono;
+import rx.RxReactiveStreams;
+
+class Scratch {
```
Comment (dmytro-lazebnyi) [inline scalecube-configuration/src/test/java/io/scalecube/configuration/CouchbaseExperimentalTest.java]: What is the purpose of having this class?
Please use `Scratch` locally without committing it to the github repository.

File: scalecube-configuration/src/main/java/io/scalecube/configuration/RequestValidator.java [hunk 8]
```diff
@@ -56,6 +66,9 @@ private static void validateRepository(String repository) {
     if (repository == null || repository.trim().isEmpty()) {
       throw new IllegalArgumentException("Please specify 'repository'");
     }
+    if (repository.contains("::")) {
```
Comment (segabriel) [inline scalecube-configuration/src/main/java/io/scalecube/configuration/RequestValidator.java:69]: Maybe should use regex, because I can specify a repository name as `bla:::bla`.

File: scalecube-configuration/src/main/java/io/scalecube/configuration/repository/couchbase/CouchbaseRepository.java [hunk 9]
```diff
@@ -38,105 +43,172 @@ public CouchbaseRepository(AsyncBucket bucket) {
   @Override
   public Mono<Boolean> createRepository(Repository repository) {
     return Mono.from(
-            RxReactiveStreams.toPublisher(
-                bucket.insert(
-                    JsonDocument.create(
-                        repository.namespace() + DELIMITER + repository.name(),
-                        JsonObject.empty()))))
+        RxReactiveStreams.toPublisher(
+            bucket.setAdd(REPOS, repository.name())
```
Comment (segabriel) [inline scalecube-configuration/src/main/java/io/scalecube/configuration/repository/couchbase/CouchbaseRepository.java]: Please apply google=format https://github.com/scalecube/scalecube-parent/blob/develop/DEVELOPMENT.md

**Review by aharonha — COMMENTED** [review]
It's quite hard to follow where is the real change in the code beneath all of the api changes.
Can you split it to two PRs?

**Comment by stingion** [comment]
API changes are in previous pull request, which are merged here. So split it again is redundant

File: scalecube-configuration/src/main/java/io/scalecube/configuration/RequestValidator.java (hunk 8 above)
Comment (stingion) [reply scalecube-configuration/src/main/java/io/scalecube/configuration/RequestValidator.java:69]: ::: contains ::

File: scalecube-configuration/src/main/java/io/scalecube/configuration/repository/couchbase/CouchbaseRepository.java (hunk 9 above)
Comment (stingion) [reply scalecube-configuration/src/main/java/io/scalecube/configuration/repository/couchbase/CouchbaseRepository.java]: Applied

File: scalecube-configuration/src/main/java/io/scalecube/configuration/repository/couchbase/CouchbaseRepository.java [hunk 10]
```diff
@@ -37,34 +42,43 @@ public CouchbaseRepository(AsyncBucket bucket) {
...
 
   @Override
-  public Mono<Document> fetch(String tenant, String repository, String key) {
+  public Mono<Document> read(String tenant, String repository, String key, Integer version) {
     return Mono.from(
             RxReactiveStreams.toPublisher(
-                bucket.mapGet(tenant + DELIMITER + repository, key, Object.class)))
+                bucket.listGet(
+                    docId(tenant, repository, key),
+                    version != null ? version - 1 : -1,
```
Comment (segabriel) [inline scalecube-configuration/src/main/java/io/scalecube/configuration/repository/couchbase/CouchbaseRepository.java:67]: -1 is it some constant? Make it more clear -> create constant for it

**Review by segabriel — APPROVED** [review]

**Review by segabriel — COMMENTED** [review]
Maybe really let's split this PR into two parts.

**Comment by stingion** [comment]
https://github.com/scalecube/scalecube-configuration-service/pull/143
